src/librustc/session/mod.rs

```diff
@@ -112,9 +112,80 @@ pub struct Session {
     /// Some measurements that are being gathered during compilation.
     pub perf_stats: PerfStats,
 
+    /// Data about code being compiled, gathered during compilation.
+    pub code_stats: RefCell<CodeStats>,
+
     next_node_id: Cell<ast::NodeId>,
 }
 
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub enum VariantSize {
+    Exact(u64),
+    Min(u64),
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct TypeSizeInfo {
+    pub type_description: String,
+    pub overall_size: u64,
+    pub variant_sizes: Option<Vec<VariantSize>>,
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct CodeStats {
+    pub type_sizes: Vec<TypeSizeInfo>,
+}
+
+impl CodeStats {
+    fn new() -> Self {
+        CodeStats { type_sizes: Vec::new() }
+    }
+
+    pub fn record_type_size<S: ToString>(&mut self,
+                                         type_desc: S,
+                                         overall_size: u64,
+                                         variant_sizes: Vec<VariantSize>) {
+        let sizes = if variant_sizes.len() == 0 { None } else { Some(variant_sizes) };
+        let info = TypeSizeInfo {
+            type_description: type_desc.to_string(),
+            overall_size: overall_size,
+            variant_sizes: sizes,
+        };
+        if !self.type_sizes.contains(&info) {
```

Comment:
Yeah true; at one time for some reason I thought there was value in preserving the order in which we encountered the types, but that's obviously not reflected in the current output.